src/librepgp/stream-parse.c

```diff
@@ -1026,22 +1006,54 @@ init_encrypted_src(pgp_processing_ctx_t *ctx, pgp_source_t *src, pgp_source_t *r
             goto finish;
         }
         if (mdcver != 1) {
-            (void) fprintf(
-              stderr, "init_encrypted_src: unknown mdc version: %d\n", (int) mdcver);
+            (void) fprintf(stderr, "init_encrypted_src: unknown mdc ver: %d\n", (int) mdcver);
             errcode = RNP_ERROR_BAD_FORMAT;
             goto finish;
         }
         param->has_mdc = true;
     }
 
     /* Obtaining the symmetric key */
+    have_key = false;
+
     if (!ctx->handler.passphrase_provider) {
         (void) fprintf(stderr, "init_encrypted_src: no passphrase provider\n");
         errcode = RNP_ERROR_BAD_PARAMETERS;
         goto finish;
     }
 
-    if (param->symencc > 0) {
+    /* Trying public-key decryption */
+    if (param->pubencc > 0) {
+        if (!ctx->handler.key_provider) {
+            (void) fprintf(stderr, "init_encrypted_src: no key provider\n");
+            errcode = RNP_ERROR_BAD_PARAMETERS;
+            goto finish;
+        }
+
+        keyctx.op = PGP_OP_DECRYPT_SYM;
+        keyctx.secret = true;
+        keyctx.stype = PGP_KEY_SEARCH_KEYID;
+
+        for (int i = 0; i < param->pubencc; i++) {
+            memcpy(keyctx.search.id, param->pubencs[i].key_id, sizeof(keyctx.search.id));
+            /* Get the key if any */
+            if (!pgp_request_key(ctx->handler.key_provider, &keyctx, &seckey)) {
+                continue;
+            }
+            /* Unlock key */
+            if (!pgp_key_unlock(seckey, ctx->handler.passphrase_provider)) {
```

Comment:
Since this isn't really a CLI-layer function, I would rather we didn't use `pgp_key_unlock`.
Normally what I would do is decrypt the seckey into a separate temporary variable and dispose of it at the end, like:
https://github.com/riboseinc/rnp/blob/master/src/lib/generate-key.c#L484

My intention for the use of `pgp_key_unlock` was always mainly FFI/API clients (and CLI would be OK too).
The general idea is:

* call `sign()` (as an example)
  * the key is already unlocked, so it is used as-is, or:
  * the key is locked, so it is temporarily decrypted and used, then disposed of

So in the case of a protected+locked key (should be the most common case), the decrypted MPIs are only exposed for a minimum window of time.

Reply:
Thanks. Looks like I mislooked the purpose of unlock or was in hurry. Will fix this.

Reply:
@dewyatt `pgp_seckey_free_secret_mpis` doesn't scrub/zero mpi contents, just calls `botan_mp_destroy`. Does it clear the contents?

Reply:
@ni4 Generally yes `botan_mp_destroy` will scrub that memory. See [here](https://botan.randombit.net/manual/side_channels.html#memory-allocation). The MP FFI interface is backed by botan's `BigInt`, which uses `secure_vector`, so that memory is scrubbed when freed.